src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Diagnostics/RazorTranslateDiagnosticsService.cs

```diff
@@ -170,15 +171,53 @@ private Diagnostic[] FilterCSharpDiagnostics(Diagnostic[] unmappedDiagnostics, R
             return false;
         }
 
-        var owner = syntaxTree.GetOwner(sourceText, d.Range.End, logger);
+        var owner = syntaxTree.FindInnermostNode(sourceText, d.Range.End, logger);
         if (owner is null)
         {
             return false;
         }
 
-        var isCSharp = owner.Kind is SyntaxKind.CSharpExpressionLiteral or SyntaxKind.CSharpStatementLiteral or SyntaxKind.CSharpEphemeralTextLiteral;
+        if (IsCsharpKind(owner))
+        {
+            return true;
+        }
+
+        // If the owner is a markup text literal, the token found
+        // could still have been the end of a csharp statement. For example:
+        // CSS123456 gets reported with the following range
+        // <style> body { overflow: [|@DateTime.Now|]; } </style>
+        // FindInnermostNode will find the encompasing MarkupTextLiteral that ends with
+        // ';' but contains the @DateTime.Now csharp expression. Check for these cases
+        // by looking to see if the start is a transition before modifying the end position
+        // to be left one.
+        if (owner.Kind is SyntaxKind.MarkupTextLiteral)
+        {
+            if (!d.Range.Start.TryGetAbsoluteIndex(sourceText, logger, out var startIndex))
+            {
+                return false;
+            }
```

Comment:
Could this be simplified by calling the overload of `FindInnermostNode` that takes a `Position`? or do we explicitly need the token here?

Reply:
alright... this is all changed